### ISSUE TITLE
helm lint misses bad chart names

### ISSUE THREAD

**Issue by mattfarina** [helm/helm#10537]
`helm lint` only catches if a chart name is missing. You can see that in the following snippet...

https://github.com/helm/helm/blob/292368a64d6c9794b4573afa5a8dc6e940a21045/pkg/lint/rules/chartfile.go#L106-L111

The `chartutil` package has the ability to validate names in some forms. We could provide better hints (like no capital letters) via lint.

Output of `helm version`: 3.7

Output of `kubectl version`: N/A

Cloud Provider/Platform (AKS, GKE, Minikube etc.): N/A

**Comment by summerlinryan** [comment]
Hey! This sounds like a great idea to me. I'd like to take this issue if possible.

**Comment by yxxhero** [comment]
@summerlinryan Do you want to work on this issue?

**Comment by summerlinryan** [comment]
@yxxhero Yes I would. To make sure I'm clear on what you were originally requesting @mattfarina you would like for `helm lint` to provide better messages for invalid chart names according to https://helm.sh/docs/chart_best_practices/conventions/#chart-names, right? For example, if someone used a dot in a chart name they might get a message like "dots are not allowed in chart names: Chart names must be lower case letters and numbers. Words may be separated with dashes."

**Comment by mattfarina** [comment]
I don't think we need to try to figure out every way someone might have incorrectly done a name. If it doesn't match the correct pattern then we can tell them the pattern it needs to fit.

**Comment by bormm** [comment]
This issue is definitely not closed.

**Comment by yxxhero** [comment]
@summerlinryan any updates?

**Comment by bormm** [comment]
> @bormm Can you create a PR for this?
I think in general the #10572 could be used, but as a very fresh helm user but also a developer I added some comments to that PR because I am not that happy with it. Maybe my comments are not correct or maybe smaftoul want to continue to work on his PR.

**Comment by Akshit42-hue** [comment]
Should I give it a try @yxxhero

**Comment by yxxhero** [comment]
@Akshit42-hue yes. You can. welcome. Anyone is welcome to contribute.

**Comment by bormm** [comment]
Before doing verification could be agree on what is a valid chart name completely?

The already mentioned https://helm.sh/docs/chart_best_practices/conventions/#chart-names contains words like "may" or "should not". The specification should not let that much room for interpretation.
Either something is allowed or not. Words that are more or less recommendation should only be used for things that could affect the performance or something. So its working, but not as efficient as possible.

Maybe the definition was not defined beforehand and is now defined by what the code can handle correctly.

Additional for the dash "-" handling: "Words may be separated with dashes (-)". As a word is not a digit or a single letter but multiple letters, it means "1-a", "1-1", "a-a", "-a" or "a-" is not allowed. Where "aa--aa" would be allowed.
But I think the description is maybe incorrect or extremely restrictive.

Additional question to #10572, as it reuses a validation function that validates with the scope for Kubernetes resources. I don't think that constrains of Kubernetes must be fulfilled for helm chart names. I am right @mattfarina ?

And I also agree with @mattfarina with the idea behind "I don't think need to try to figure out every way someone might have incorrectly done a name". So don't make a special error message for every invalid character. But I disagree with his conclusion for the "can tell them the pattern it needs to fit".
Helm messages are not (only) for machines. I think giving messages that are optimized to be machine readable like a regular expression should not be the goal. 

I think a compromises would be just print a part of the the @summerlinryan suggested "Chart names must be lower case letters and numbers. Words may be separated with dashes." in all invalid cases. But of course as mentioned above, it should be very clear without the "may".
We could also print out a URL to a webpage containing the definition.

**Comment by piroux** [comment]
> As a word is not a digit or a single letter but multiple letters, it means "1-a", "1-1", "a-a", "-a" or "a-" is not allowed. Where "aa--aa" would be allowed.

Yes, it is a bit extreme. :smile: 
In many human languages, there are words with only one letter. How about this instead :

Allowed: "aa--aa", "1-a", "1-1", "a-a"
Not allowed: "-a",  "a-"

What do you all think ?

**Comment by Moynur** [comment]
hey this is still waiting for a review if anyone has some time https://github.com/helm/helm/pull/11298

**Comment by robertsirc** [comment]
@Moynur the branch is out dated with conflicts, might be easier for a review without conflicts.

**Comment by Moynur** [comment]
sure I'll rebase it been hanging for a while

**Comment by Moynur** [comment]
@robertsirc should be good again

**Comment by JanitSri** [comment]
@Moynur are you still working on this? If not, I can try to tackle it.

**Comment by Moynur** [comment]
Feel free to, happy to rebase my branch if someone wants to review it too
